### PR TITLE
Update to formatters V2 and add durations

### LINKED ISSUE #2: Use `itemDuration` to add duration
Hspec `Formatters.V2` has `itemDuration` that can be used to add test case/group duration.

https://www.stackage.org/haddock/lts-20.10/hspec-core-2.9.7/Test-Hspec-Core-Formatters-V2.html#g:2

### PR DESCRIPTION
**Pull request by hobovsky**
Solves https://github.com/codewars/hspec-formatters-codewars/issues/2

### CODE REVIEW

**Comment by hobovsky** [comment]
I tried to compose assertion messages in a way as close to default HSpec reporter as possible:

image

image

File: src/Test/Hspec/Formatters/Codewars.hs [hunk 1]
```diff
@@ -1,50 +1,73 @@
...
 codewars =
   silent
-    { exampleGroupStarted = \_ name -> do
+    {
+      formatterGroupStarted = \path -> do
         writeLine ""
-        writeLine $ escapeLF $ "<DESCRIBE::>" ++ name
-    , exampleGroupDone = writeLine "\n<COMPLETEDIN::>"
-    , exampleSucceeded = \(_, name) _ -> do
+        writeLine $ escapeLF $ "<DESCRIBE::>" ++ (getName path)
```
Comment (kazk) [inline src/Test/Hspec/Formatters/Codewars.hs]:
```suggestion
        writeLine $ "<DESCRIBE::>" ++ (getName path)
```
LF is handled by `getName`.

File: src/Test/Hspec/Formatters/Codewars.hs [hunk 2]
```diff
@@ -1,50 +1,73 @@
...
         writeLine ""
-        writeLine $ escapeLF $ "<IT::>" ++ name
-        writeLine "\n<PASSED::>Test Passed"
-        writeLine "\n<COMPLETEDIN::>"
-    , exampleFailed = \(_, name) _ reason -> do
+        writeLine $ "<COMPLETEDIN::>"
+      ,formatterItemStarted = \path -> do
         writeLine ""
-        writeLine $ escapeLF $ "<IT::>" ++ name
+        writeLine $ escapeLF $ "<IT::>" ++ (getName path)
```
Comment (kazk) [inline src/Test/Hspec/Formatters/Codewars.hs]:
```suggestion
        writeLine $ "<IT::>" ++ (getName path)
```

File: src/Test/Hspec/Formatters/Codewars.hs [hunk 3]
```diff
@@ -1,50 +1,73 @@
...
+      ,formatterItemDone = \_ item -> do
         writeLine ""
-        writeLine $ escapeLF $ reasonAsString reason
-        writeLine "\n<COMPLETEDIN::>"
+        writeLine $ reportItem item
+        writeLine ""
+        writeLine $ "<COMPLETEDIN::>" ++ (formatSeconds $ itemDuration item)
```
Comment (kazk) [inline src/Test/Hspec/Formatters/Codewars.hs]: `<COMPLETEDIN::>` should have milliseconds.

File: test/Main.hs [hunk 4]
```diff
@@ -1,10 +1,25 @@
+{-# LANGUAGE OverloadedStrings #-}
+
 module Main (main) where
 
+import Data.Text (pack, replace, unpack)
+import System.Exit
+
 import Test.Hspec.Runner
 import Test.Hspec.Formatters.Codewars (codewars)
```
Comment (kazk) [inline test/Main.hs]: We can export `escapeLF` (maybe improve the name) and import it here.

Comment (hobovsky) [reply test/Main.hs]: I do not know :(

Do you mean to create a new module (i.e. a `*.hs` file), move the function there, and reuse it in two places?

Comment (kazk) [reply test/Main.hs]: Export it from `codewars.hs`
```haskell
module Test.Hspec.Formatters.Codewars (codewars, escapeLF) where
```
then import it here

```haskell
import Test.Hspec.Formatters.Codewars (codewars, escapeLF)
```

Comment (kazk) [reply test/Main.hs]: ~~`LF` in `escapeLF` looks kind of awkward, so maybe rename to `escapeLf` or `escapeLn`?~~

Never mind, `LF` should be fine.

File: src/Test/Hspec/Formatters/Codewars.hs (hunk 1 above)
Comment (kazk) [reply src/Test/Hspec/Formatters/Codewars.hs]: `escapeLF` came back.

Comment (hobovsky) [reply src/Test/Hspec/Formatters/Codewars.hs]: Copy/paste mistake again.

File: test/Main.hs [hunk 5]
```diff
@@ -1,10 +1,23 @@
+{-# LANGUAGE OverloadedStrings #-}
```
Comment (kazk) [inline test/Main.hs]: Do we still need this?

File: src/Test/Hspec/Formatters/Codewars.hs [hunk 6]
```diff
@@ -1,50 +1,81 @@
...
+      ,formatterGroupDone = \_ -> do
         writeLine ""
-        writeLine $ escapeLF $ "<DESCRIBE::>" ++ name
-    , exampleGroupDone = writeLine "\n<COMPLETEDIN::>"
-    , exampleSucceeded = \(_, name) _ -> do
+        ts <- liftIO $ readIORef times
+        now <- getRealTime
+        let startedOn = head ts
+        let duration = now - startedOn
+        writeLine $ "<COMPLETEDIN::> " ++ (formatToMillis $ duration)
```
Comment (kazk) [inline src/Test/Hspec/Formatters/Codewars.hs]:
```suggestion
        writeLine $ "<COMPLETEDIN::>" ++ (formatToMillis $ duration)
```

File: test/Main.hs [hunk 7]
```diff
@@ -1,10 +1,23 @@
...
+
 module Main (main) where
 
+import Data.Text (pack, replace, unpack)
```
Comment (kazk) [inline test/Main.hs]:
```suggestion
```

File: test/Main.hs (hunk 5 above)
Comment (hobovsky) [reply test/Main.hs]: No, I forgot to remove it after removing `escapeLF`.

Deleted.

**Review by kazk — APPROVED** [review]

**Pull request merged**